File: r/tests/test_sparse_dense_equivalence.R
#!/usr/bin/env Rscript
# Test that sparse and dense matrix implementations produce identical results
#
# This verifies that our sparse matrix optimization doesn't change the numerical
# results compared to a dense matrix implementation.

library(polars)
library(Matrix)
source("R/common.R")
source("R/compress.R")

set.seed(42)

cat("Testing sparse vs dense matrix equivalence...\n\n")

# Test 1: Coefficients are identical
cat("Test 1: Sparse vs dense coefficients identical... ")

n <- 1000
df <- pl$DataFrame(
  y = rnorm(n, 10, 2),
  x1 = rnorm(n),
  x2 = rnorm(n, 5, 2),
  fe1 = rep(0:49, each = 20),
  fe2 = rep(0:19, 50)
)

# Compress data
comp_result <- .compress_polars(df, "y", c("x1", "x2"), c("fe1", "fe2"))
compressed <- comp_result$compressed

# Build design matrix - sparse
dm_sparse <- .build_design_matrix(compressed, c("x1", "x2"), c("fe1", "fe2"), use_sparse = TRUE)
X_sparse <- dm_sparse$X
Y <- dm_sparse$Y
wts <- dm_sparse$wts

# Build design matrix - dense
dm_dense <- .build_design_matrix(compressed, c("x1", "x2"), c("fe1", "fe2"), use_sparse = FALSE)
X_dense <- dm_dense$X

# Solve WLS
wls_sparse <- .solve_wls(X_sparse, Y, wts)
wls_dense <- .solve_wls(X_dense, Y, wts)

# Compare coefficients (first 2 are x1, x2)
beta_sparse <- wls_sparse$beta[1:2]
beta_dense <- wls_dense$beta[1:2]

stopifnot(all(abs(beta_sparse - beta_dense) < 1e-10))
cat("PASSED\n")

# Test 2: XtX_inv is identical
cat("Test 2: Sparse vs dense XtX_inv identical... ")

XtX_inv_sparse <- wls_sparse$XtX_inv
XtX_inv_dense <- wls_dense$XtX_inv

stopifnot(all(abs(XtX_inv_sparse - XtX_inv_dense) < 1e-10))
cat("PASSED\n")

# Test 3: RSS is identical
cat("Test 3: Sparse vs dense RSS identical... ")

rss_sparse <- .compute_rss_grouped(compressed, X_sparse, wls_sparse$beta)
rss_dense <- .compute_rss_grouped(compressed, X_dense, wls_dense$beta)

stopifnot(abs(rss_sparse$rss_total - rss_dense$rss_total) < 1e-8)
stopifnot(all(abs(rss_sparse$rss_g - rss_dense$rss_g) < 1e-8))
cat("PASSED\n")

# Test 4: IID SEs are identical
cat("Test 4: Sparse vs dense IID SEs identical... ")

n_obs <- comp_result$n_obs_original
df_resid <- n_obs - length(dm_sparse$all_col_names)

se_iid_sparse <- .compute_se_compress(
  XtX_inv_sparse, rss_sparse$rss_total, rss_sparse$rss_g,
  n_obs, df_resid, "iid", X_sparse, 2
)

se_iid_dense <- .compute_se_compress(
  XtX_inv_dense, rss_dense$rss_total, rss_dense$rss_g,
  n_obs, df_resid, "iid", X_dense, 2
)

stopifnot(all(abs(se_iid_sparse$se - se_iid_dense$se) < 1e-10))
cat("PASSED\n")

# Test 5: HC1 SEs are identical
cat("Test 5: Sparse vs dense HC1 SEs identical... ")

se_hc1_sparse <- .compute_se_compress(
  XtX_inv_sparse, rss_sparse$rss_total, rss_sparse$rss_g,
  n_obs, df_resid, "HC1", X_sparse, 2
)

se_hc1_dense <- .compute_se_compress(
  XtX_inv_dense, rss_dense$rss_total, rss_dense$rss_g,
  n_obs, df_resid, "HC1", X_dense, 2
)

stopifnot(all(abs(se_hc1_sparse$se - se_hc1_dense$se) < 1e-10))
cat("PASSED\n")

# Test 6: Cluster SEs are identical
cat("Test 6: Sparse vs dense cluster SEs identical... ")

# Create data with clusters
df_cluster <- pl$DataFrame(
  y = rnorm(600, 10, 2),
  x = rnorm(600),
  fe1 = rep(0:29, each = 20),
  cluster = rep(0:29, each = 20)
)

comp_cluster <- .compress_polars(df_cluster, "y", "x", "fe1", cluster_col = "cluster")
compressed_c <- comp_cluster$compressed

dm_sparse_c <- .build_design_matrix(compressed_c, "x", "fe1", use_sparse = TRUE)
dm_dense_c <- .build_design_matrix(compressed_c, "x", "fe1", use_sparse = FALSE)

wls_sparse_c <- .solve_wls(dm_sparse_c$X, dm_sparse_c$Y, dm_sparse_c$wts)
wls_dense_c <- .solve_wls(dm_dense_c$X, dm_dense_c$Y, dm_dense_c$wts)

rss_sparse_c <- .compute_rss_grouped(compressed_c, dm_sparse_c$X, wls_sparse_c$beta)
rss_dense_c <- .compute_rss_grouped(compressed_c, dm_dense_c$X, wls_dense_c$beta)

# Get cluster IDs and e0_g
pdf <- as.data.frame(compressed_c)
cluster_ids <- pdf[["cluster"]]
n_g <- pdf[["_n"]]
sum_y_g <- pdf[["_sum_y"]]
yhat_sparse <- as.vector(dm_sparse_c$X %*% wls_sparse_c$beta)
yhat_dense <- as.vector(dm_dense_c$X %*% wls_dense_c$beta)
e0_g_sparse <- sum_y_g - n_g * yhat_sparse
e0_g_dense <- sum_y_g - n_g * yhat_dense

n_obs_c <- comp_cluster$n_obs_original
df_resid_c <- n_obs_c - length(dm_sparse_c$all_col_names)

se_cluster_sparse <- .compute_se_compress(
  wls_sparse_c$XtX_inv, rss_sparse_c$rss_total, rss_sparse_c$rss_g,
  n_obs_c, df_resid_c, "cluster", dm_sparse_c$X, 1,
  cluster_ids = cluster_ids, e0_g = e0_g_sparse
)

se_cluster_dense <- .compute_se_compress(
  wls_dense_c$XtX_inv, rss_dense_c$rss_total, rss_dense_c$rss_g,
  n_obs_c, df_resid_c, "cluster", dm_dense_c$X, 1,
  cluster_ids = cluster_ids, e0_g = e0_g_dense
)

stopifnot(se_cluster_sparse$n_clusters == se_cluster_dense$n_clusters)
stopifnot(all(abs(se_cluster_sparse$se - se_cluster_dense$se) < 1e-10))
cat("PASSED\n")

cat("\n✅ All 6 sparse vs dense equivalence tests passed!\n")

File: python/tests/test_sparse_dense_equivalence.py
"""Test that sparse and dense matrix implementations produce identical results."""
import polars as pl
import numpy as np
import pytest
from leanfe.compress import (
    build_design_matrix,
    solve_wls,
    compute_rss_grouped,
    compute_se_compress,
    compress_polars,
)


def test_sparse_dense_coefficients_identical():
    """Verify sparse and dense implementations produce bit-identical coefficients."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'y': np.random.normal(10, 2, n),
        'x1': np.random.normal(0, 1, n),
        'x2': np.random.normal(0, 1, n),
        'fe1': np.random.randint(0, 50, n),
        'fe2': np.random.randint(0, 20, n),
    })
    
    # Compress data
    compressed, n_obs = compress_polars(df, 'y', ['x1', 'x2'], ['fe1', 'fe2'])
    
    # Build design matrix - sparse
    X_sparse, Y, wts, cols_sparse, _ = build_design_matrix(
        compressed, ['x1', 'x2'], ['fe1', 'fe2'], backend='polars', use_sparse=True
    )
    
    # Build design matrix - dense
    X_dense, Y_dense, wts_dense, cols_dense, _ = build_design_matrix(
        compressed, ['x1', 'x2'], ['fe1', 'fe2'], backend='polars', use_sparse=False
    )
    
    # Solve WLS
    beta_sparse, XtX_inv_sparse = solve_wls(X_sparse, Y, wts)
    beta_dense, XtX_inv_dense = solve_wls(X_dense, Y_dense, wts_dense)
    
    # Coefficients should be nearly identical (within numerical precision)
    np.testing.assert_allclose(beta_sparse[:2], beta_dense[:2], rtol=1e-10)
    
    # XtX_inv should also match
    np.testing.assert_allclose(XtX_inv_sparse, XtX_inv_dense, rtol=1e-10)


def test_sparse_dense_se_identical():
    """Verify sparse and dense SE computations produce identical results."""
    np.random.seed(123)
    n = 500
    
    df = pl.DataFrame({
        'y': np.random.normal(10, 2, n),
        'x': np.random.normal(0, 1, n),
        'fe1': np.random.randint(0, 30, n),
        'cluster': np.random.randint(0, 20, n),
    })
    
    # Compress with cluster
    compressed, n_obs = compress_polars(df, 'y', ['x'], ['fe1'], cluster_col='cluster')
    
    # Build design matrices
    X_sparse, Y, wts, cols, _ = build_design_matrix(
        compressed, ['x'], ['fe1'], backend='polars', use_sparse=True
    )
    X_dense, Y_d, wts_d, cols_d, _ = build_design_matrix(
        compressed, ['x'], ['fe1'], backend='polars', use_sparse=False
    )
    
    # Solve WLS
    beta_sparse, XtX_inv_sparse = solve_wls(X_sparse, Y, wts)
    beta_dense, XtX_inv_dense = solve_wls(X_dense, Y_d, wts_d)
    
    # Compute RSS
    rss_total_s, rss_g_s = compute_rss_grouped(compressed, X_sparse, beta_sparse, 'polars')
    rss_total_d, rss_g_d = compute_rss_grouped(compressed, X_dense, beta_dense, 'polars')
    
    # RSS should match (allow small numerical differences from floating point)
    np.testing.assert_allclose(rss_total_s, rss_total_d, rtol=1e-7)
    np.testing.assert_allclose(rss_g_s, rss_g_d, rtol=1e-7)
    
    # Test IID SEs
    df_resid = n_obs - len(cols)
    se_iid_s, _ = compute_se_compress(
        XtX_inv_sparse, rss_total_s, rss_g_s, n_obs, df_resid, 'iid', X_sparse, ['x']
    )
    se_iid_d, _ = compute_se_compress(
        XtX_inv_dense, rss_total_d, rss_g_d, n_obs, df_resid, 'iid', X_dense, ['x']
    )
    np.testing.assert_allclose(se_iid_s, se_iid_d, rtol=1e-10)
    
    # Test HC1 SEs
    se_hc1_s, _ = compute_se_compress(
        XtX_inv_sparse, rss_total_s, rss_g_s, n_obs, df_resid, 'HC1', X_sparse, ['x']
    )
    se_hc1_d, _ = compute_se_compress(
        XtX_inv_dense, rss_total_d, rss_g_d, n_obs, df_resid, 'HC1', X_dense, ['x']
    )
    np.testing.assert_allclose(se_hc1_s, se_hc1_d, rtol=1e-10)


def test_sparse_dense_cluster_se_identical():
    """Verify sparse and dense cluster SE computations produce identical results."""
    np.random.seed(456)
    n = 600
    
    df = pl.DataFrame({
        'y': np.random.normal(10, 2, n),
        'x': np.random.normal(0, 1, n),
        'fe1': np.random.randint(0, 30, n),
        'cluster': np.repeat(np.arange(30), 20),  # 30 clusters, 20 obs each
    })
    
    # Compress with cluster
    compressed, n_obs = compress_polars(df, 'y', ['x'], ['fe1'], cluster_col='cluster')
    
    # Build design matrices
    X_sparse, Y, wts, cols, _ = build_design_matrix(
        compressed, ['x'], ['fe1'], backend='polars', use_sparse=True
    )
    X_dense, Y_d, wts_d, cols_d, _ = build_design_matrix(
        compressed, ['x'], ['fe1'], backend='polars', use_sparse=False
    )
    
    # Solve WLS
    beta_sparse, XtX_inv_sparse = solve_wls(X_sparse, Y, wts)
    beta_dense, XtX_inv_dense = solve_wls(X_dense, Y_d, wts_d)
    
    # Compute RSS
    rss_total_s, rss_g_s = compute_rss_grouped(compressed, X_sparse, beta_sparse, 'polars')
    rss_total_d, rss_g_d = compute_rss_grouped(compressed, X_dense, beta_dense, 'polars')
    
    # Get cluster IDs and e0_g
    cluster_ids = compressed['cluster'].to_numpy()
    n_g = compressed['_n'].to_numpy()
    sum_y_g = compressed['_sum_y'].to_numpy()
    yhat_g_s = np.asarray(X_sparse @ beta_sparse).flatten()
    yhat_g_d = X_dense @ beta_dense
    e0_g_s = sum_y_g - n_g * yhat_g_s
    e0_g_d = sum_y_g - n_g * yhat_g_d
    
    df_resid = n_obs - len(cols)
    
    # Test cluster SEs
    se_cluster_s, n_clusters_s = compute_se_compress(
        XtX_inv_sparse, rss_total_s, rss_g_s, n_obs, df_resid, 'cluster', X_sparse, ['x'],
        cluster_ids=cluster_ids, e0_g=e0_g_s
    )
    se_cluster_d, n_clusters_d = compute_se_compress(
        XtX_inv_dense, rss_total_d, rss_g_d, n_obs, df_resid, 'cluster', X_dense, ['x'],
        cluster_ids=cluster_ids, e0_g=e0_g_d
    )
    
    np.testing.assert_equal(n_clusters_s, n_clusters_d)
    np.testing.assert_allclose(se_cluster_s, se_cluster_d, rtol=1e-10)
